shared/idmap/shift_linux.go

```diff
@@ -280,18 +280,63 @@ static int get_userns_fd_cb(void *data)
 static int get_userns_fd(void)
 {
 	int ret;
+	int fd = -EBADF;
 	pid_t pid;
 	char path[256];
 
+	// Create the namespace.
 	pid = do_clone(get_userns_fd_cb, NULL, CLONE_NEWUSER);
 	if (pid < 0)
 		return -errno;
 
+	// Fetch a reference.
 	snprintf(path, sizeof(path), "/proc/%d/ns/user", pid);
 	ret = open(path, O_RDONLY | O_CLOEXEC);
+
+	// Kill the temporary process.
 	kill(pid, SIGKILL);
 	wait_for_pid(pid);
-	return ret;
+
+	if (ret != 0)
+		return ret;
+
+	// Setup uid_map
+	snprintf(path, sizeof(path), "/proc/%d/uid_map", pid);
+	fd = openat(AT_FDCWD, path, O_WRONLY);
+	if (fd < 0)
+		return -errno;
+
+	if (write(fd, "0 0 1", 5) != 5)
+		return -errno;
+
+	if (close(fd) < 0)
+		return -errno;
+
+	// Setup setgroups
+	snprintf(path, sizeof(path), "/proc/%d/setgroups", pid);
+	fd = openat(AT_FDCWD, path, O_WRONLY);
+	if (fd < 0)
+		return -errno;
+
+	if (write(fd, "deny", 4) != 4)
+		return -errno;
+
+	if (close(fd) < 0)
+		return -errno;
+
+	// Setup gid_map
+	snprintf(path, sizeof(path), "/proc/%d/gid_map", pid);
+	fd = openat(AT_FDCWD, path, O_WRONLY);
+	if (fd < 0)
+		return -errno;
+
+	if (write(fd, "0 0 1", 5) != 5)
+		return -errno;
```

Comment:
Note that since this is c and not go, fd will not be cleaned up in this error path.

Reply:
Fixed!